### ISSUE TITLE
Statements not cleaned up

### ISSUE THREAD

**Issue by davidmartos96** [simolus3/drift#1917]
Hello Simon, I've recently encountered a race condition when using sharing a database file between a desktop Flutter app and a python script with the sqlite3 module.
The flutter app only reads while the python script may insert.

What ends up happening sometimes is that a drift select throws a database is locked error. This is momentary and consecutive reads in drift work fine.
The real problem is that after this error, the connection in the python script is not able to write to the database, the drift app needs to be fully closed.
Taking a look at the code I was able to find that inside the runSelect function, a prepared statement is created, then executed and then disposed.
If an error happens in the execute step, the prepared statement is not disposed. I tried disposing it regardless with a finally clause and this issue doesn't happen anymore.
There are 3 places were I think this could happen. 
runSelect, runBatched and runWithArgs

I tried creating a Dart test but I wasn't able to. A Busy error when doing a database read when not using exclusive locks is very rare, plus this is a multiprocess setup.

**Comment by davidmartos96** [comment]
@simolus3  I made a small reproduction error, both a dart and python script

To reproduce:
1. `python main.py`
2. In a separate terminal, `dart run drift_app.dart`

This will make the drift app (sqlite reader) start showing Database is locked error, and at the same time, the python script (sqlite writer) will stop working, even when restarting it.
Following reads in the dart script keep working.

After changing the implmentation of `runSelect` in https://github.com/simolus3/drift/blob/6049d1aaa019623617b05164f27ae689bbed68d5/drift/lib/src/sqlite3/database.dart#L138

into the following, it starts working fine (locked errors still appear, but the writer keeps working
```dart
@override
  Future<QueryResult> runSelect(String statement, List<Object?> args) async {
    final stmt = _db.prepare(statement, checkNoTail: true);

    try {
      final result = stmt.select(args);
      return Future.value(QueryResult.fromRows(result.toList()));
    } finally {
      stmt.dispose();
    }
  }
```
----
## Scripts

Python script: `main.py`
```python
import random
import sqlite3
from time import sleep


def main():
    conn = sqlite3.connect('test.db')

    conn.execute('''
        CREATE TABLE IF NOT EXISTS
        users (id INTEGER PRIMARY KEY, name TEXT)
    ''')
    conn.commit()


    while True:
        c = conn.cursor()
        random_num = random.randint(1, 100)
        print("Inserting {}".format(random_num))
        c.execute('INSERT OR REPLACE INTO users(id, name) VALUES(?, ?)', (1, 'some_name' + str(random_num)))
        conn.commit()
        sleep(1)

main()
```

Dart script: `drift_app.dart`
```dart
import 'dart:async';
import 'dart:io';

import 'package:drift/drift.dart';
import 'package:drift/native.dart';

void main(List<String> args) async {
  final db = NativeDatabase(File("test.db"));
  await db.ensureOpen(_FakeExecutorUser());

  print("Start periodically reading. Only errors will get logged.");
  Timer.periodic(Duration(milliseconds: 50), (timer) async {
    try {
      final res = await db.runSelect('''
SELECT *
FROM users
''', []);

      final mappedCards = res.map((row) => row["name"]).toList();
      //print(mappedCards);
    } catch (e) {
      print("Something went wrong: $e");
    }
  });
}

class _FakeExecutorUser extends QueryExecutorUser {
  @override
  Future<void> beforeOpen(QueryExecutor executor, OpeningDetails details) {
    return Future.value();
  }

  @override
  int get schemaVersion => 1;
}
```

**Comment by simolus3** [comment]
Thanks for the helpful issue! I agree that the `dispose()` call needs to be moved into a `finally` block and for me that also solves the issue with the python script. I'll try to find a repro that can be turned into a unit test more easily.

**Comment by simolus3** [comment]
I've fixed this in 626ef29eff6e4c5d34f74844d40090f0d1af5e0d. I couldn't find a good way to turn this into a Dart-only test unfortunately (I tried to come up with something that essentially runs the Dart-equivalent of your Python program in a background isolate, but that was still way too flaky).

I've also used some neat new Dart features to automatically dispose statements and databases when they're no longer referenced in https://github.com/simolus3/sqlite3.dart/commit/4df5e1bd4d084304f92078e1cd07538e1038b1e0.

**Comment by davidmartos96** [comment]
> I've fixed this in 626ef29eff6e4c5d34f74844d40090f0d1af5e0d. I couldn't find a good way to turn this into a Dart-only test unfortunately (I tried to come up with something that essentially runs the Dart-equivalent of your Python program in a background isolate, but that was still way too flaky).
> 
> I've also used some neat new Dart features to automatically dispose statements and databases when they're no longer referenced in https://github.com/simolus3/sqlite3.dart/commit/4df5e1bd4d084304f92078e1cd07538e1038b1e0.

Yes, unfortunately a unit test here is quite tricky.

The finalizer functionality looks nice! When do they run, when the Garbage Collector triggers and no one has a hold to the object?

**Comment by simolus3** [comment]
Published the fix into the 1.7.1 hotfix version.

**Comment by davidmartos96** [comment]
> Published the fix into the 1.7.1 hotfix version.

Thank you!
